herd/AArch64Sem.ml

```diff
@@ -2033,6 +2183,187 @@ module Make
         reduce (nelem-1) (M.add (V.intToV 0) (V.intToV 0)) >>=
           fun v -> write_reg_neon_sz sz r1 v ii
 
+      let uaddv var r1 p src ii =
+        let nelem = predicate_nelem src in
+        let psize = predicate_psize src in
+        let esize = scalable_esize src in
+        read_reg_predicate false p ii >>= fun pred ->
+          get_predicate_any pred psize nelem >>= fun any ->
+            M.choiceT
+            any
+            (read_reg_scalable false src ii >>= fun src ->
+              let read_active_elem_or_zero idx cur_val =
+                get_predicate_last pred psize idx >>= fun last ->
+                  M.choiceT
+                  last
+                  (scalable_getlane cur_val idx esize)
+                  mzero
+                in
+                let rec reduce cur_val n op =
+                match n with
+                | 0 -> op >>| read_active_elem_or_zero n cur_val >>= fun (v1,v2) -> M.add v1 v2
+                | _ -> reduce cur_val (n-1) (op >>| read_active_elem_or_zero n cur_val  >>= fun (v1,v2) -> M.add v1 v2)
+                in
+                reduce src (nelem-1) mzero)
+            mzero >>= fun v ->
+              let sz = AArch64Base.tr_simd_variant var in
+              write_reg_neon_sz sz r1 v ii
+
+      let movprfx dst pg src ii =
+        let nelem = predicate_nelem src in
+        let psize = predicate_psize src in
+        let esize = scalable_esize dst in
+        let orig = match pg with
+        | AArch64Base.PMreg (_,AArch64Base.Zero)
+          -> mzero
+        | AArch64Base.PMreg (_,AArch64Base.Merge)
+          -> read_reg_scalable false dst ii
+        | _ -> assert false
+        in
+        orig >>|
+        read_reg_predicate false pg ii >>= fun (orig,pred) ->
+          get_predicate_any pred psize nelem >>= fun any ->
```

Comment:
That's an interesting implementation choice.

Your implementation generates one Register Read/Write effect per register. Another approach would be to generate one Register Read/Write effect per lane.

A first observation is that if the predicate is \M and all, then you don't have to read the dst register. But more importantly, the pseudocode makes it look like these register effects are per lane.

What do you think?

Reply:
Ignore this, I misread (again) the pseudocode. I think your code is fine!